### PR TITLE
added application time tracker

### PR DESCRIPTION
**Pull request by Hyjale**

### CODE REVIEW

**Comment by charmoniumQ** [comment]
- [x] Knock out `cpu_timer.hpp` (and remove from `CMakeLists..txt`. This was part of a failed experiment that only works in C++.
- [x] If we still need `ts`, `ts2`, `ts3`, `ts4` (ask @mhuzai ), then replace with `start_timer/stop_timer/print_timer` (see `t5` for example). Otherwise, just remove the lines referencing these timers.

```C
// Bad
struct timespec ts;
clock_gettime(CLOCK_REALTIME, &ts);
/* ... */
struct timespec ts1;
clock_gettime(CLOCK_REALTIME, &ts1);
unsigned long long duration = (ts1.tv_sec - ts.tv_sec) * 1000000000 + ts1.tv_nsec - ts.tv_nsec;
printf("cpu_timer,application_wall,%llu\n", duration);
```

```C
// Good
timer t5 = {.name = "application_inner",.clock_id = CLOCK_REALTIME,};
start_timer(&t5);
/* ... */
stop_timer(&t5);
print_timer(&t5);
```

**Comment by mhuzai** [comment]
Why are there so many changes in addition to the timers? For instance, to the OpenGL shaders, the blitting code, and the OpenXR logic itself. It's also really hard to tell because of the indentation change.

Re: timers. We do not need `ts1 - ts` (wall time) anymore. However, we do still need `ts4 - ts3` (CPU time). But replace the latter with what Sam said (`timer t = {...`).

**Comment by charmoniumQ** [comment]
@mhuzai , That bothered me too until [this](https://stackoverflow.com/questions/37007300/how-to-ignore-whitespace-in-github-when-comparing).

File: main.c [hunk 1]
```diff
@@ -759,33 +923,35 @@ main_loop(xr_example* self)
...
 
-		// --- Wait for our turn to render a frame
-		XrFrameState frameState = {.type = XR_TYPE_FRAME_STATE, .next = NULL};
-		XrFrameWaitInfo frameWaitInfo = {.type = XR_TYPE_FRAME_WAIT_INFO,
-		                                 .next = NULL};
-		result = xrWaitFrame(self->session, &frameWaitInfo, &frameState);
-		if (!xr_result(self->instance, result,
-		               "xrWaitFrame() was not successful, exiting..."))
-			break;
+		timer ts = {.name = "application_cpu",.clock_id = CLOCK_THREAD_CPUTIME_ID};
```
Comment (mhuzai) [inline main.c]:
```suggestion
		timer tCPU = {.name = "application_cpu",.clock_id = CLOCK_THREAD_CPUTIME_ID};
```

File: main.c [hunk 2]
```diff
@@ -759,33 +923,35 @@ main_loop(xr_example* self)
...
+		start_timer(&ts);
```
Comment (mhuzai) [inline main.c]:
```suggestion
		start_timer(&tCPU);
```

File: main.c [hunk 3]
```diff
@@ -1033,9 +1273,8 @@ main_loop(xr_example* self)
 		if (!xr_result(self->instance, result, "failed to end frame!"))
 			break;
 
-		blitNvsync(self->configuration_views[1].recommendedImageRectWidth,
-			            self->configuration_views[1].recommendedImageRectHeight,
-			            self->framebuffers[1][bufferIndex]);
+		stop_timer(&ts);
```
Comment (mhuzai) [inline main.c]:
```suggestion
		stop_timer(&tCPU);
```

File: main.c [hunk 4]
```diff
@@ -1033,9 +1273,8 @@ main_loop(xr_example* self)
...
+		print_timer(&ts);
```
Comment (mhuzai) [inline main.c]:
```suggestion
		print_timer(&tCPU);
```

File: main.c [hunk 5]
```diff
@@ -1002,13 +1237,18 @@ main_loop(xr_example* self)
 			projection_views[i].subImage.imageRect.extent.height =
 			    self->configuration_views[i].recommendedImageRectHeight;
 
+			timer t5 = {.name = "application_inner",.clock_id = CLOCK_REALTIME,};
```
Comment (mhuzai) [inline main.c]:
```suggestion
			timer tWall = {.name = "application_inner",.clock_id = CLOCK_REALTIME,};
```

File: main.c [hunk 6]
```diff
@@ -1002,13 +1237,18 @@ main_loop(xr_example* self)
...
+			start_timer(&t5);
```
Comment (mhuzai) [inline main.c]:
```suggestion
			start_timer(&tWall);
```

File: main.c [hunk 7]
```diff
@@ -1002,13 +1237,18 @@ main_loop(xr_example* self)
...
+
 			renderFrame(self->configuration_views[i].recommendedImageRectWidth,
 			            self->configuration_views[i].recommendedImageRectHeight,
 			            projectionMatrix, inverseViewMatrix, &spaceLocation[0], &spaceLocation[1],
 			            self->framebuffers[i][bufferIndex], self->depthbuffer,
 			            self->images[i][bufferIndex], i,
 			            frameState.predictedDisplayTime);
 			glFinish();
+			stop_timer(&t5);
```
Comment (mhuzai) [inline main.c]:
```suggestion
			stop_timer(&tWall);
```

File: main.c [hunk 8]
```diff
@@ -1002,13 +1237,18 @@ main_loop(xr_example* self)
...
+			print_timer(&t5);
```
Comment (mhuzai) [inline main.c]:
```suggestion
			print_timer(&tWall);
```

File: main.c [hunk 9]
```diff
@@ -22,12 +22,57 @@
...
+	assert(!r);
+
+	timer->diff = timespec_sub(&timer->stop, &timer->start);
+}
+
+static void print_timer(timer* timer) {
+	assert(timer->name);
+	int r = printf("cpu_timer,%s,%llu\n", timer->name, timer->diff);
+	assert(r);
+}
```
Comment (mhuzai) [inline main.c]: @charmoniumQ should this code be in its own header file for readability?

Comment (charmoniumQ) [reply main.c]: Yeah, I suppose it could be.

Comment (mhuzai) [reply main.c]: Okay. @Hyjale please move it to its own header file then.

**Review by mhuzai — APPROVED** [review]

**Review by charmoniumQ — APPROVED** [review]

**Pull request merged**